### PR TITLE
Add tests for HttpListenerContext.AcceptWebSocketAsync and fix the managed implementation

### PR DESCRIPTION
**Pull request by hughbe**
First commit fixes duplicate code and adds a missing EditorBrowsable attribute to a method

Second commit cleans up platform independent code for validating the arguments to AcceptWebSocketAsync in preparation for fixing the managed implementation

Third commit adds a bunch of tests for functionality related to this method (they passed before this PR). It then fixes the tests for the managed implementation. Previously almost all of these tests failed with either debug assertitions, incorrect exceptions thrown or no exception thrown.

Contributes to #18128

@stephentoub @priya91 @viktorhofer

### CODE REVIEW

File: src/System.Net.HttpListener/tests/HttpListenerContextTests.cs [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+                HttpListenerWebSocketContext socketContext = context.AcceptWebSocketAsync(serverProtocol).Result;
+                Assert.Equal(serverProtocol, socketContext.WebSocket.SubProtocol);
+            }, subProtocols: clientProtocols);
+        }
+
+        [ConditionalFact(nameof(IsNotWindows7))]
+        // The managed implementation doesn't validate that the socket is actually a web socket.
+        // This means that HttpListener can connect to a Socket masquerading as a web socket.
+        [ActiveIssue(18128, TestPlatforms.AnyUnix)]
+        public async Task AcceptWebSocketAsync_SocketSpoofingAsWebSocket_ThrowsWebSocketException()
```
Comment (hughbe) [inline src/System.Net.HttpListener/tests/HttpListenerContextTests.cs]: This is odd - basically we can spoof HttpListener into connecting to a WebSocket that isn't actually a WebSocket by sending the right headers. Obviously this is a bug, I'm don't know if it's a security issue though

File: src/System.Net.HttpListener/tests/HttpListenerContextTests.cs [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+            yield return new object[] { new string[] { "MyProtocol1", "MyProtocol2" }, "MyProtocol2" };
+        }
+
+        [ConditionalTheory(nameof(IsNotWindows7))]
+        [MemberData(nameof(SubProtocol_TestData))]
+        public async Task AcceptWebSocketAsync_ValidSubProtocol_Success(string[] clientProtocols, string serverProtocol)
+        {
+            await GetWebSocketContext((socket, context) =>
+            {
+                HttpListenerWebSocketContext socketContext = context.AcceptWebSocketAsync(serverProtocol).Result;
```
Comment (ViktorHofer) [inline src/System.Net.HttpListener/tests/HttpListenerContextTests.cs]: Any reason why you await the Result synchronously? You could make the Action async: `async (socket, context) => ...`

File: src/System.Net.HttpListener/tests/HttpListenerContextTests.cs [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+        [InlineData("random=text")]
+        [InlineData("random{text")]
+        [InlineData("random}text")]
+        [InlineData("\x19")]
+        [InlineData("\x7f")]
+        public async Task AcceptWebSocketAsync_InvalidSubProtocol_ThrowsArgumentException(string subProtocol)
+        {
+            await GetWebSocketContext((socket, context) =>
+            {
+                Assert.ThrowsAsync<ArgumentException>("subProtocol", () => context.AcceptWebSocketAsync(subProtocol)).Wait();
```
Comment (ViktorHofer) [inline src/System.Net.HttpListener/tests/HttpListenerContextTests.cs]: Inside here you could also add async to the Action and await `content.AcceptWebSocketAsync(...)`. Same for the outer scope and ThrowsAsync

Comment (ViktorHofer) [inline src/System.Net.HttpListener/tests/HttpListenerContextTests.cs]: Not sure because I haven't cloned your branch but you should get compiler warnings because you aren't awaiting the async method.

File: src/System.Net.HttpListener/tests/HttpListenerContextTests.cs [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+        [InlineData("Connection: Upgrade\r\nUpgrade: websocket\r\nSec-WebSocket-Version: ")]
+        [InlineData("Connection: Upgrade\r\nUpgrade: websocket\r\nSec-WebSocket-Version: 1")]
+        [InlineData("Connection: Upgrade\r\nUpgrade: websocket\r\nSec-WebSocket-Version: 13")]
+        [InlineData("Connection: Upgrade\r\nUpgrade: websocket\r\nSec-WebSocket-Version: 13\r\nSec-WebSocket-Key: ")]
+        [InlineData("UnknownHeader: random")]
+        public async Task AcceptWebSocketAsync_InvalidHeaders_ThrowsWebSocketException(string headers)
+        {
+            await GetSocketContext(new string[] { headers }, context =>
+            {
+                Assert.ThrowsAsync<WebSocketException>(() => context.AcceptWebSocketAsync(null)).Wait();
```
Comment (ViktorHofer) [inline src/System.Net.HttpListener/tests/HttpListenerContextTests.cs]: Same here. I'll stop commenting on these.

File: src/System.Net.HttpListener/src/System/Net/WebSockets/HttpWebSocket.cs [hunk 5]
```diff
@@ -81,6 +82,96 @@ internal static string GetSecWebSocketAcceptString(string secWebSocketKey)
...
+        }
+
+        internal const int MinSendBufferSize = 16;
+        internal const int MinReceiveBufferSize = 256;
+        internal const int MaxBufferSize = 64 * 1024;
+
+        private static void ValidateWebSocketHeaders(HttpListenerContext context)
+        {
+            ValidateWebSocketHeadersCore(context);
+
```
Comment (ViktorHofer) [inline src/System.Net.HttpListener/src/System/Net/WebSockets/HttpWebSocket.cs]: nit: empty line

File: src/System.Net.HttpListener/tests/HttpListenerContextTests.cs (hunk 2 above)
Comment (hughbe) [inline src/System.Net.HttpListener/tests/HttpListenerContextTests.cs]: I found that the tests hang if an assertion in the async lambda fails, but I was probably doing something wrong. I just pushed a commit that cleans up the tests to avoid using lambdas/callbacks altogether

File: src/System.Net.HttpListener/tests/HttpListenerContextTests.cs (hunk 1 above)
Comment (stephentoub) [inline src/System.Net.HttpListener/tests/HttpListenerContextTests.cs]: I don't understand the "masquerading as a web socket" part.  Web sockets are just layered on top of http, which are implemented on sockets, so every "web socket" is a socket.  It sounds like you're saying that the managed implementation simply isn't validating all of the headers the way it should, in which case that's something we should fix.

File: src/System.Net.HttpListener/src/System/Net/Managed/WebSockets/HttpWebSocket.Managed.cs [hunk 6]
```diff
@@ -75,45 +77,7 @@ internal static partial class HttpWebSocket
...
-
-            if (string.IsNullOrWhiteSpace(context.Request.Headers[HttpKnownHeaderNames.SecWebSocketKey]))
-            {
-                throw new WebSocketException(WebSocketError.HeaderError,
-                    SR.Format(SR.net_WebSockets_AcceptHeaderNotFound,
-                    nameof(ValidateWebSocketHeaders),
-                    HttpKnownHeaderNames.SecWebSocketKey));
-            }
-        }
+        private static void ValidateWebSocketHeadersCore(HttpListenerContext context) { }
```
Comment (stephentoub) [inline src/System.Net.HttpListener/src/System/Net/Managed/WebSockets/HttpWebSocket.Managed.cs]: You could also do this as a partial method, in which case you won't need an empty method like this in the managed implementation.

File: src/System.Net.HttpListener/src/System/Net/WebSockets/HttpWebSocket.cs (hunk 5 above)
Comment (stephentoub) [inline src/System.Net.HttpListener/src/System/Net/WebSockets/HttpWebSocket.cs]: This empty line is fine.  There are empty lines between each piece of the validation.

File: src/System.Net.HttpListener/src/System/Net/Windows/WebSockets/HttpWebSocket.Windows.cs [hunk 7]
```diff
@@ -18,20 +18,17 @@ internal static partial class HttpWebSocket
         private static readonly Random s_keyGenerator = new Random();
         private static readonly bool s_httpSysSupportsWebSockets = (Environment.OSVersion.Version >= new Version(6, 2));
 
-        internal static ArraySegment<byte> EmptyPayload
-        {
-            get { return s_EmptyPayload; }
```
Comment (stephentoub) [inline src/System.Net.HttpListener/src/System/Net/Windows/WebSockets/HttpWebSocket.Windows.cs]: Nit: `new ArraySegment<byte>(new byte[] { }, 0, 0)` above can be `new ArraySegment<byte>(Array.Empty<byte>())`

File: src/System.Net.HttpListener/src/System/Net/Windows/WebSockets/HttpWebSocket.Windows.cs [hunk 8]
```diff
@@ -320,35 +221,13 @@ internal static void ThrowPlatformNotSupportedException_WSPC()
...
-            }
-            if (arraySegment.Count < 0 || arraySegment.Count > (arraySegment.Array.Length - arraySegment.Offset))
-            {
-                throw new ArgumentOutOfRangeException(parameterName + "." + nameof(arraySegment.Count));
-            }
-        }
+        private static string SupportedVersion => WebSocketProtocolComponent.SupportedVersion;
 
-        private static void EnsureHttpSysSupportsWebSockets()
+        private static void ValidateWebSocketHeadersCore(HttpListenerContext context)
```
Comment (stephentoub) [inline src/System.Net.HttpListener/src/System/Net/Windows/WebSockets/HttpWebSocket.Windows.cs]: Rather than calling this ValidateWebSocketHeadersCore, how about ThrowIfWebSocketsNotSupported?

Comment (stephentoub) [inline src/System.Net.HttpListener/src/System/Net/Windows/WebSockets/HttpWebSocket.Windows.cs]: Or alternatively have a WebSocketsSupported property, and have the PNSE throw be at the call site: then the managed implementation would always return true and the Windows would return s_httpSysSupportsWebSockets.

File: src/System.Net.HttpListener/tests/HttpListenerContextTests.cs [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+
+        [ConditionalFact(nameof(IsNotWindows7OrUapCore))]
+        // The managed implementation doesn't validate that the socket is actually a web socket.
+        // This means that HttpListener can connect to a Socket masquerading as a web socket.
+        [ActiveIssue(18128, TestPlatforms.AnyUnix)]
+        public async Task AcceptWebSocketAsync_SocketSpoofingAsWebSocket_ThrowsWebSocketException()
+        {
+            await GetSocketContext(new string[] { "Connection: Upgrade", "Upgrade: websocket", "Sec-WebSocket-Version: 13", "Sec-WebSocket-Key: Key" }, context =>
+            {
+                Assert.ThrowsAsync<WebSocketException>(() => context.AcceptWebSocketAsync(null)).Wait();
```
Comment (stephentoub) [inline src/System.Net.HttpListener/tests/HttpListenerContextTests.cs]: Blocking like this with Wait is problematic.  The problem is that xunit uses a synchronization context that only allows for N pieces of test code to be running at the same time.  Let's say N == 1.  If that test blocks waiting for an async operation to complete, and that async operation itself has queued some work back to the SynchronizationContext to be run (the default with `await task;`), then the only thread that could possibly execute that continuation is blocked waiting for that continuation, and you deadlock.

The right pattern here is to change GetSocketContext to take a `Func<context, Task>` instead of an `Action<context>`.  Then you'd do:
```C#
async context =>
{
    await Assert.ThrowsAsync<WebSocketException>(() => context.AcceptWebSocketAsync(null));
]
```

File: src/System.Net.HttpListener/tests/HttpListenerContextTests.cs [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+                foreach (string subProtocol in subProtocols)
+                {
+                    Socket.Options.AddSubProtocol(subProtocol);
+                }
+            }
+
+            var uriBuilder = new UriBuilder(Factory.ListeningUrl) { Scheme = "ws" };
+            Task<HttpListenerContext> serverContextTask = Factory.GetListener().GetContextAsync();
+
+            Task clientConnectTask = Socket.ConnectAsync(uriBuilder.Uri, CancellationToken.None);
```
Comment (stephentoub) [inline src/System.Net.HttpListener/tests/HttpListenerContextTests.cs]: This this connection fails, the test will end up hanging, even though we could have accurately reported the failure that was stored in the clientConnectTask.  That's what was happening in the test I fixed here: https://github.com/dotnet/corefx/pull/19755

File: src/System.Net.HttpListener/tests/HttpListenerContextTests.cs [hunk 11]
```diff
@@ -0,0 +1,178 @@
...
+using System.Threading;
+using System.Threading.Tasks;
+using Xunit;
+
+namespace System.Net.Tests
+{
+    public class HttpListenerContextTests : IDisposable
+    {
+        private HttpListenerFactory Factory { get; }
+        private ClientWebSocket Socket { get; }
```
Comment (stephentoub) [inline src/System.Net.HttpListener/tests/HttpListenerContextTests.cs]: Where is this used?  It's a good idea to have this and to have tests that use it to connect, and validate that we can actually connect and have a web socket conversation between the client and server, but I don't see such tests here.

File: src/System.Net.HttpListener/tests/HttpListenerContextTests.cs (hunk 1 above)
Comment (hughbe) [inline src/System.Net.HttpListener/tests/HttpListenerContextTests.cs]: I assumed WebSockets and Sockets were different entities despite the name. The error that we get in the Windows implementation occurs when sending the headers to the socket:

> An error occurred when sending the WebSocket HTTP upgrade response during the AcceptWebSocketAsync operation. The HRESULT returned is '1229'

Comment (stephentoub) [inline src/System.Net.HttpListener/tests/HttpListenerContextTests.cs]: You can see a simple web socket client connection logic here:
https://github.com/dotnet/corefx/blob/master/src/System.Net.WebSockets.Client/src/System/Net/WebSockets/WebSocketHandle.Managed.cs#L74
It just opens a socket to the target, and sends an HTTP request with the relevant headers, then validates the response headers, at which point the web socket connection is considered opened.  If the Windows implementation is failing and the managed succeeding here, that suggests that the HttpListener validation is different between the two, and we should understand and fix the differences in the managed one.

Comment (hughbe) [inline src/System.Net.HttpListener/tests/HttpListenerContextTests.cs]: Mind if I do that in a separate PR? I want to consolidate more non-argument validation logic in `AcceptWebSocketAsync` but things might get a bit messy and a bit longer/trickier to review when combined with all the other changes in this PR

File: src/System.Net.HttpListener/src/System/Net/Windows/WebSockets/HttpWebSocket.Windows.cs (hunk 7 above)
Comment (hughbe) [inline src/System.Net.HttpListener/src/System/Net/Windows/WebSockets/HttpWebSocket.Windows.cs]: I went with `ArraySegment<byte>.Empty` and folded calls to EmptyPayload into it's caller

File: src/System.Net.HttpListener/src/System/Net/WebSockets/HttpWebSocket.cs (hunk 5 above)
Comment (hughbe) [inline src/System.Net.HttpListener/src/System/Net/WebSockets/HttpWebSocket.cs]: This is now an if-block so the new line helps :)

File: src/System.Net.HttpListener/src/System/Net/Windows/WebSockets/HttpWebSocket.Windows.cs (hunk 8 above)
Comment (hughbe) [inline src/System.Net.HttpListener/src/System/Net/Windows/WebSockets/HttpWebSocket.Windows.cs]: That's a lot nicer, done

File: src/System.Net.HttpListener/tests/HttpListenerContextTests.cs (hunk 11 above)
Comment (hughbe) [inline src/System.Net.HttpListener/tests/HttpListenerContextTests.cs]: I was planning on adding a bunch of WebSocket tests for functionality after this PR. For now it is used to create a context and is also disposed below.

There are also some tests in WebSocketTests.cs that perform a conversation - they're pretty minimal so I was hoping to expand on them

File: src/System.Net.HttpListener/tests/HttpListenerContextTests.cs (hunk 9 above)
Comment (hughbe) [inline src/System.Net.HttpListener/tests/HttpListenerContextTests.cs]: Interesting, I went with that approach, thanks

File: src/System.Net.HttpListener/tests/HttpListenerContextTests.cs (hunk 10 above)
Comment (hughbe) [inline src/System.Net.HttpListener/tests/HttpListenerContextTests.cs]: Fixed in PR feedback

File: src/System.Net.HttpListener/tests/HttpListenerContextTests.cs (hunk 1 above)
Comment (stephentoub) [inline src/System.Net.HttpListener/tests/HttpListenerContextTests.cs]: That's fine.

File: src/System.Net.HttpListener/tests/HttpListenerContextTests.cs (hunk 11 above)
Comment (stephentoub) [inline src/System.Net.HttpListener/tests/HttpListenerContextTests.cs]: Ok, thanks.

File: src/System.Net.HttpListener/src/System/Net/Windows/WebSockets/HttpWebSocket.Windows.cs [hunk 12]
```diff
@@ -252,60 +203,6 @@ internal static void ValidateInnerStream(Stream innerStream)
...
-            if (keepAliveInterval < Timeout.InfiniteTimeSpan) // -1
-            {
-                throw new ArgumentOutOfRangeException(nameof(keepAliveInterval), keepAliveInterval,
-                    SR.Format(SR.net_WebSockets_ArgumentOutOfRange_TooSmall, Timeout.InfiniteTimeSpan.ToString()));
-            }
-        }
-
-        internal static void ValidateBufferSizes(int receiveBufferSize, int sendBufferSize)
-        {
-            if (receiveBufferSize < WebSocketBuffer.MinReceiveBufferSize)
```
Comment (stephentoub) [inline src/System.Net.HttpListener/src/System/Net/Windows/WebSockets/HttpWebSocket.Windows.cs]: Looks like we're removing a bunch of references to members of WebSocketBuffer.  Can any of those members be removed, or do they all still have other references to them?

Comment (hughbe) [inline src/System.Net.HttpListener/src/System/Net/Windows/WebSockets/HttpWebSocket.Windows.cs]: I removed them here in the second commit: https://github.com/dotnet/corefx/pull/19761/commits/f7078b76e92c23576d735e5dab8ab5f2f5c45083#diff-3f8f0d1e604c4e521e7c1f7dc83cd197L28

**Comment by stephentoub** [comment]
@dotnet-bot help

**Comment by stephentoub** [comment]
@dotnet-bot test outerloop netcoreapp Windows 7 Debug
@dotnet-bot test outerloop netcoreapp Windows 10 Debug
@dotnet-bot test outerloop netcoreapp Ubuntu14.04 Debug

**Comment by stephentoub** [comment]
@dotnet-bot test outerloop netcoreapp Windows_NT Debug

File: src/Common/src/System/Net/WebSockets/WebSocketValidate.cs [hunk 13]
```diff
@@ -123,9 +124,19 @@ internal static void ThrowPlatformNotSupportedException()
...
         internal static void ValidateArraySegment(ArraySegment<byte> arraySegment, string parameterName)
         {
+            Debug.Assert(!string.IsNullOrEmpty(parameterName), "'parameterName' MUST NOT be NULL or string.Empty");
+
             if (arraySegment.Array == null)
             {
-                throw new ArgumentNullException(parameterName + ".Array");
+                throw new ArgumentNullException(parameterName + "." + nameof(arraySegment.Array));
+            }
+            if (arraySegment.Offset < 0 || arraySegment.Offset > arraySegment.Array.Length)
```
Comment (CIPop) [inline src/Common/src/System/Net/WebSockets/WebSocketValidate.cs]: Do these match the .Net Framework behavior?

File: src/Common/src/System/Net/WebSockets/WebSocketValidate.cs [hunk 14]
```diff
@@ -123,9 +124,19 @@ internal static void ThrowPlatformNotSupportedException()
 
         internal static void ValidateArraySegment(ArraySegment<byte> arraySegment, string parameterName)
```
Comment (CIPop) [inline src/Common/src/System/Net/WebSockets/WebSocketValidate.cs]: This change require additional tests for the WebSockets contract. Will this class behave similar to .Net Framework? What about WebSocket.Client?

**Comment by stephentoub** [comment]
@dotnet-bot test OuterLoop netcoreapp Windows 10 Debug x64 please

File: src/System.Net.HttpListener/tests/HttpListenerContextTests.cs [hunk 15]
```diff
@@ -0,0 +1,184 @@
...
+
+            yield return new object[] { new string[] { "MyProtocol1" }, null };
+            yield return new object[] { new string[] { "MyProtocol" }, "MyProtocol" };
+            yield return new object[] { new string[] { "MyProtocol" }, "myPROTOCOL" };
+
+            yield return new object[] { new string[] { "MyProtocol1", "MyProtocol2" }, null };
+            yield return new object[] { new string[] { "MyProtocol1", "MyProtocol2" }, "MyProtocol2" };
+        }
+
+        [ConditionalTheory(nameof(IsNotWindows7OrUapCore))]
```
Comment (CIPop) [inline src/System.Net.HttpListener/tests/HttpListenerContextTests.cs]: Nit: UAP core should remain as an active issue rather than a condition to ensure we get to enable them during the UAP work.

File: src/System.Net.HttpListener/src/System/Net/Windows/WebSockets/HttpWebSocket.Windows.cs [hunk 16]
```diff
@@ -14,24 +14,17 @@ namespace System.Net.WebSockets
...
         internal static Task<HttpListenerWebSocketContext> AcceptWebSocketAsync(HttpListenerContext context,
             string subProtocol,
             int receiveBufferSize,
             TimeSpan keepAliveInterval,
             ArraySegment<byte> internalBuffer)
         {
-            ValidateOptions(subProtocol, receiveBufferSize, WebSocketBuffer.MinSendBufferSize, keepAliveInterval);
-            ValidateArraySegment(internalBuffer, nameof(internalBuffer));
-            WebSocketBuffer.Validate(internalBuffer.Count, receiveBufferSize, WebSocketBuffer.MinSendBufferSize, true);
+            ValidateOptions(subProtocol, receiveBufferSize, MinSendBufferSize, keepAliveInterval);
```
Comment (CIPop) [inline src/System.Net.HttpListener/src/System/Net/Windows/WebSockets/HttpWebSocket.Windows.cs]: I'm not sure I understand this change: why are we changing from the .Net Framework port to unify with a new, managed implementation.
Is there an assumption here that the buffer bounds are the same between using the HTTP.SYS APIs and the managed one? (In this case it may - which would make the change OK - just confirming.)

File: src/Common/src/System/Net/WebSockets/WebSocketValidate.cs (hunk 13 above)
Comment (hughbe) [inline src/Common/src/System/Net/WebSockets/WebSocketValidate.cs]: I wasn't sure about how to add tests for this change, as I don't know how to create this scenario.

Previously, the managed implementation used this method, but the windows implementation used a method with these additional checks.

By the way, is this dead code? I don't know how it's possible for these extra checks to be true? Surely argument validation in ArraySegment (https://github.com/dotnet/coreclr/blob/40ecbb723552407044b57edaf3e5a1408bf99aab/src/mscorlib/src/System/ArraySegment.cs) would remove the need for this?

@stephentoub  what do you think?

File: src/System.Net.HttpListener/src/System/Net/Windows/WebSockets/HttpWebSocket.Windows.cs (hunk 16 above)
Comment (hughbe) [inline src/System.Net.HttpListener/src/System/Net/Windows/WebSockets/HttpWebSocket.Windows.cs]: > I'm not sure I understand this change: why are we changing from the .Net Framework port to unify with a new, managed implementation.

We are unifying the managed and windows implementation to fix managed bugs. In all these situations, as you will see, we prefer the Windows code. So in essence, what I've done is move non-Windows specific code to the PAL layer, removed the managed implementation code that does the same thing - thus Windows and managed have the same behaviour by definition as they have the same code.

> Is there an assumption here that the buffer bounds are the same between using the HTTP.SYS APIs and the managed one? (In this case it may - which would make the change OK - just confirming.)

Previously some of the inputs in the unit tests I've added would hit Debug assertions as they are invalid in the managed stream.

Basically we want almost complete Windows compatability in the managed implementation. I believe one of the reasons for this is that we plan to use the managed implementation of HttpListener in UWP on Windows.

File: src/Common/src/System/Net/WebSockets/WebSocketValidate.cs (hunk 13 above)
Comment (hughbe) [inline src/Common/src/System/Net/WebSockets/WebSocketValidate.cs]: I tried getting to this situation with:
```cs
internal static void ValidateArraySegment(ArraySegment<byte> arraySegment, string parameterName)
{
    if (arraySegment.Array == null)
    {
        throw new ArgumentNullException(parameterName + "." + nameof(arraySegment.Array));
    }
    if (arraySegment.Offset < 0 || arraySegment.Offset > arraySegment.Array.Length)
    {
        throw new ArgumentOutOfRangeException(parameterName + "." + nameof(arraySegment.Offset));
    }
    if (arraySegment.Count < 0 || arraySegment.Count > (arraySegment.Array.Length - arraySegment.Offset))
    {
        throw new ArgumentOutOfRangeException(parameterName + "." + nameof(arraySegment.Count));
    }
}

static void Main(string[] args)
{
    byte[] array = new byte[10];

    ArraySegment<byte> arraySegment = new ArraySegment<byte>(array, 0, 10);
    ValidateArraySegment(arraySegment, nameof(array));

    Array.Resize(ref array, 1);
    ValidateArraySegment(arraySegment, nameof(array));
}
```

but it seems like ArraySegment holds on the old array, not the sized array internal

File: src/System.Net.HttpListener/tests/HttpListenerContextTests.cs (hunk 15 above)
Comment (hughbe) [inline src/System.Net.HttpListener/tests/HttpListenerContextTests.cs]: Really? If you look at all the other tests, they are all conditional theories. This is because UAP core doesn't have httpapi.dll i think

File: src/Common/src/System/Net/WebSockets/WebSocketValidate.cs (hunk 13 above)
Comment (hughbe) [inline src/Common/src/System/Net/WebSockets/WebSocketValidate.cs]: Even nasty reflection seems not to work:
```cs
byte[] array = new byte[10];

ArraySegment<byte> arraySegment = new ArraySegment<byte>(array, 0, 10);

FieldInfo field = arraySegment.GetType().GetField("_offset", BindingFlags.NonPublic | BindingFlags.Instance);
field.SetValue(arraySegment, -1);
Console.WriteLine(field.GetValue(arraySegment)); // Prints 0!
```

Comment (stephentoub) [inline src/Common/src/System/Net/WebSockets/WebSocketValidate.cs]: That's why these checks were previously removed.  Though it is possible.  See the trick employed here: https://github.com/dotnet/corefx/blob/773b2573afc047717d1366f21a9c3fb0d9e05f51/src/System.Net.Sockets/tests/FunctionalTests/SendReceive.cs#L52

Comment (hughbe) [inline src/Common/src/System/Net/WebSockets/WebSocketValidate.cs]: That is wonderfully sneaky and somewhat scary - do you have a suggestion of what to do with the checks in this PR - should I add them to WebSocketValidate.ValidateArraySegment (could break .net compat) or just add my own validator to HttpWebSocket in HttpListener.

Obviously I'll add tests for any changes

Comment (stephentoub) [inline src/Common/src/System/Net/WebSockets/WebSocketValidate.cs]: Desktop's WebSocketValidate.ValidateArraySegment looks like this:
```
        internal static void ValidateArraySegment<T>(ArraySegment<T> arraySegment, string parameterName)
        {
            Contract.Requires(!string.IsNullOrEmpty(parameterName), "'parameterName' MUST NOT be NULL or string.Empty");

            if (arraySegment.Array == null)
            {
                throw new ArgumentNullException(parameterName + ".Array");
            }

            if (arraySegment.Offset < 0 || arraySegment.Offset > arraySegment.Array.Length)
            {
                throw new ArgumentOutOfRangeException(parameterName + ".Offset");
            }
            if (arraySegment.Count < 0 || arraySegment.Count > (arraySegment.Array.Length - arraySegment.Offset))
            {
                throw new ArgumentOutOfRangeException(parameterName + ".Count");
            }
        }
```
so I wouldn't be concerned about breaking compat with it.  But it's also not clear it's necessary.  Such validation is important if the data is then passed to something that, for example, manipulates the input using unsafe code such that "sneaky" code like I shared could get invalid and potentially malicious into the unsafe code unvalidated.

**Comment by hughbe** [comment]
Test Innerloop Ubuntu16.04 arm Debug Cross Build

**Comment by stephentoub** [comment]
@dotnet-bot test outerloop netfx Windows_NT Debug
@dotnet-bot test outerloop netcoreapp Windows_NT Debug
@dotnet-bot test outerloop netcoreapp Ubuntu14.04 Debug

File: src/System.Net.HttpListener/tests/HttpListenerContextTests.cs (hunk 15 above)
Comment (CIPop) [inline src/System.Net.HttpListener/tests/HttpListenerContextTests.cs]: Then it's fine to leave them as-is. We'll change them later.
